### PR TITLE
Switch to ESM, change node requirement to node 14

### PR DESCRIPTION
**Pull request by void-spark**

### CODE REVIEW

File: index.mjs [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
+const token = '%[a-f0-9]{2}';
+const singleMatcher = new RegExp('(' + token + ')|([^%]+?)', 'gi');
+const multiMatcher = new RegExp('(' + token + ')+', 'gi');
 
 function decodeComponents(components, split) {
 	try {
 		// Try to decode the entire string first
 		return [decodeURIComponent(components.join(''))];
-	} catch (err) {
+	} catch (error) {
```
Comment (SamVerschueren) [inline index.mjs:11]: Remove the `(error)` binding here. I believe it's not needed anymore.

Comment (SamVerschueren) [reply index.mjs:11]: If we do that everywhere, we can remove the disabled `unicorn/prefer-optional-catch-binding`.

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -9,15 +9,11 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        node-version:
+        node-version:        
+          - '19'
```
Comment (SamVerschueren) [inline .github/workflows/main.yaml:13]: Remove `19`. I only care about LTS (just like `query-string`)

File: test.mjs [hunk 3]
```diff
@@ -21,7 +21,7 @@ const tests = {
 	'%ab': '%ab',
 	'%ab%ab%ab': '%ab%ab%ab',
 	'%61+%4d%4D': 'a MM',
-	'\uFEFFtest': '\uFEFFtest',
+	'\uFEFFtest2': '\uFEFFtest2',
```
Comment (SamVerschueren) [inline test.mjs:24]: Any reason this was updated?

File: test.mjs [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import test from 'ava';
-import m from './';
+import m from './index.mjs';
```
Comment (SamVerschueren) [inline test.mjs:2]: Rename the test file back to `test.js`.

**Review by SamVerschueren — CHANGES_REQUESTED** [review]
Gave some feedback. I think overall this is pretty good 👍 .

Can you also update the `readme.md` to use `import decodeURIComponent from 'decode-uri-component';` instead of the require?

File: test.mjs (hunk 3 above)
Comment (void-spark) [reply test.mjs:24]: It caused the same test title twice ('test -> test'), which isn't allowed with the new test libraries.

File: package.json [hunk 5]
```diff
@@ -10,14 +10,18 @@
...
 		"test": "xo && nyc ava",
 		"coveralls": "nyc report --reporter=text-lcov | coveralls"
 	},
+	"type": "module",
+	"exports": {
+		"default": "./index.mjs"
+	},
 	"files": [
-		"index.js"
+		"index.mjs"
```
[the remark this answers is not in the document]
Comment (void-spark) [reply package.json]: Done

File: test.mjs (hunk 4 above)
Comment (void-spark) [reply test.mjs]: Done

File: .github/workflows/main.yaml (hunk 2 above)
Comment (void-spark) [reply .github/workflows/main.yaml]: Done

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "decode-uri-component",
-	"version": "0.2.2",
+	"version": "0.3.0",
```
[the remark this answers is not in the document]
Comment (void-spark) [reply package.json]: Done

File: package.json [hunk 7]
```diff
@@ -10,14 +10,18 @@
...
-		"node": ">=0.10"
+		"node": ">=14.16"
 	},
 	"scripts": {
 		"test": "xo && nyc ava",
 		"coveralls": "nyc report --reporter=text-lcov | coveralls"
 	},
+	"type": "module",
+	"exports": {
+		"default": "./index.mjs"
```
[the remark this answers is not in the document]
Comment (void-spark) [reply package.json]: Done

File: index.mjs [hunk 8]
```diff
@@ -87,8 +87,8 @@ module.exports = function (encodedURI) {
 
 		// Try the built in decoder first
 		return decodeURIComponent(encodedURI);
-	} catch (err) {
+	} catch (error) {
```
[the remark this answers is not in the document]
Comment (void-spark) [reply index.mjs]: done

File: index.mjs [hunk 9]
```diff
@@ -42,18 +43,18 @@ function decode(input) {
...
 
-	var match = multiMatcher.exec(input);
+	let match = multiMatcher.exec(input);
 	while (match) {
 		try {
 			// Decode as big chunks as possible
 			replaceMap[match[0]] = decodeURIComponent(match[0]);
-		} catch (err) {
-			var result = decode(match[0]);
+		} catch (error) {
```
[the remark this answers is not in the document]
Comment (void-spark) [reply index.mjs]: done

File: index.mjs [hunk 10]
```diff
@@ -18,19 +19,19 @@ function decodeComponents(components, split) {
...
 
 	return Array.prototype.concat.call([], decodeComponents(left), decodeComponents(right));
 }
 
 function decode(input) {
 	try {
 		return decodeURIComponent(input);
-	} catch (err) {
-		var tokens = input.match(singleMatcher) || [];
+	} catch (error) {
```
[the remark this answers is not in the document]
Comment (void-spark) [reply index.mjs]: done

File: index.mjs (hunk 1 above)
Comment (void-spark) [reply index.mjs]: done and done

File: index.mjs [hunk 11]
```diff
@@ -1,13 +1,14 @@
-'use strict';
-var token = '%[a-f0-9]{2}';
-var singleMatcher = new RegExp('(' + token + ')|([^%]+?)', 'gi');
-var multiMatcher = new RegExp('(' + token + ')+', 'gi');
+/* eslint-disable unicorn/prefer-optional-catch-binding,no-unused-vars,import/no-anonymous-default-export */
```
[the remark this answers is not in the document]
Comment (void-spark) [reply index.mjs]: The (error), which is now gone :)

File: index.mjs [hunk 12]
```diff
@@ -66,18 +67,17 @@ function customDecodeURIComponent(input) {
...
 		// Replace all decoded components
-		var key = entries[i];
 		input = input.replace(new RegExp(key, 'g'), replaceMap[key]);
 	}
 
 	return input;
 }
 
-module.exports = function (encodedURI) {
+export default function (encodedURI) {
```
[the remark this answers is not in the document]
Comment (void-spark) [reply index.mjs]: Yep, and done.
Only I used 'decodeUriComponent' instead of the existing/system 'decodeURIComponent', which explodes the test.

**Review by SamVerschueren — APPROVED** [review]

**Comment by SamVerschueren** [comment]
Awesome work @void-spark 🙏 !

**Comment by SamVerschueren** [comment]
Released as 0.4.0.

**Comment by void-spark** [comment]
Well, now @sindresorhus just needs to update the dependencies to the ECM versions:
    "decode-uri-component": "^0.4.0",
    "split-on-first": "^3.0.0"
Seems anyone who hasn't committed to query-string cannot open or comment on issues anymore, so hope he sees this :)

**Comment by SamVerschueren** [comment]
Just sent a PR to the `query-string` package with the updated versions.

**Review by zezo2019 — APPROVED** [review]

**Comment by HAWKS3CURE** [comment]
void-spark:main.patch:x

**Comment by sandritis2** [comment]
14

**Pull request merged**